### PR TITLE
Keras add layer

### PR DESCRIPTION
**Pull request by tprocter46**
Following from Issue #170 , an implementation of the keras "add" layer - just adds (two or more) same-sized outputs together.

Also includes a test - so there's an imminent PR to the test-data repo incoming.
At the moment, the `$INPUT` variable in the test points to the branch of my fork of the test-data repo to make it work, but I can change this depending on which way round it's normal to submit PRs.

### CODE REVIEW

File: tests/test-Add.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#  - Read in the saved file, with C++, test on a few patterns
+#  - Compare the outputs to somehting previously saved
+
+# _______________________________________________________________________
+# define inputs, outputs, and code to run
+#
+# If to add another test you'll probably have to edit this
+
+# Trained network to convert and test
+INPUT=https://github.com/tprocter46/lwtnn-test-data/raw/add-test/add_dense_functional.tgz
```
Comment (dguest) [inline tests/test-Add.sh:16]: I merged your pull request in the other repository and tagged it as v10, so this can point there now.

**Review by dguest — COMMENTED** [review]
Overall looks pretty great, thanks. A few minor comments.

File: tests/test-Add.sh (hunk 1 above)
Comment (tprocter46) [reply tests/test-Add.sh]: Thanks! Updated.

File: include/lwtnn/generic/Graph.tcc [hunk 2]
```diff
@@ -152,6 +152,41 @@ namespace generic {
...
 
+  template<typename T>
+  AddNode<T>::AddNode(const std::vector<const INode<T>*>& sources):
+    m_sources(sources)
+  {
+    assert(sources.size() > 0);
+    m_n_outputs = sources[0]->n_outputs();
+    //Check to make sure each input layer has the same size
+    for (const auto source: sources) {
+      assert(source->n_outputs() == m_n_outputs);
```
[the remark this answers is not in the document]
Comment (tprocter46) [reply include/lwtnn/generic/Graph.tcc]: I was trying to copy the style from the ConcatenateNode compute method, but that explanation makes sense. Thanks - useful to know!
I missed the first one originally, so it's across two commits, sorry - but fixed.

[the remark this answers is not in the document]
Comment (dguest) [reply include/lwtnn/generic/Graph.tcc]: Yeah we're using `assert` in the `compute` methods a bit, where it makes sense to strip them out.

File: include/lwtnn/generic/Graph.tcc [hunk 3]
```diff
@@ -469,7 +508,15 @@ namespace generic {
...
       m_nodes[iii] = new SumNode<T>(m_seq_nodes.at(node.sources.at(0)));
-    } else {
+    } else if (node.type == NodeConfig::Type::ADD) {
+      // build add layer
+      std::vector<const INode<T>*> in_nodes;
+      for (std::size_t source_node: node.sources) {
+        in_nodes.push_back(m_nodes.at(source_node));
+      }
+      m_nodes[iii] = new AddNode<T>(in_nodes);
+    }
```
Comment (dguest) [inline include/lwtnn/generic/Graph.tcc:518]: The linter seems quite upset about this line: it has a space after the `}` and we don't allow trailing whitespace.

**Review by dguest — COMMENTED** [review]
Everything seems alright, but you've gone and made the linter very angry, better fix that.

File: include/lwtnn/generic/Graph.tcc [hunk 4]
```diff
@@ -152,6 +152,45 @@ namespace generic {
...
+    for (const auto source: sources) {
+      if(source->n_outputs() != m_n_outputs){
+        throw NNConfigurationException("All sources of an add layer must have same dimension");
+      }
+    }
+  }
+
+  template<typename T>
+  VectorX<T> AddNode<T>::compute(const ISource<T>& source) const {
+    VectorX<T> output(m_n_outputs);
```
[the remark this answers is not in the document]
Comment (tprocter46) [reply include/lwtnn/generic/Graph.tcc]: `I haven't done that much work directly interfacing with Eigen, so I didn't know about the `VectorX::Zero` - thanks - it's definitely cleaner than before, and halves the length of the function.

I also used the opportunity to move the input.rows() call inside the assert - it probably won't make much difference, but if the assert gets stripped out then there's no point assigning an unused `size_t`, right? (though I guess the compiler might  realise this and won't actually call it anyway).

File: include/lwtnn/generic/Graph.tcc (hunk 3 above)
Comment (tprocter46) [reply include/lwtnn/generic/Graph.tcc]: Fixed

**Pull request merged**